### ISSUE TITLE
sh performance and streaming behavior

### ISSUE THREAD

**Issue by fivetentaylor** [amoffat/sh#235]
# Introduction

First off, I just want to compliment you on how rad this module is!!!

I really think this concept is transformational to how people use python.

I noticed when I started to use sh that things ran a little slower than a comparable shell job, or even a python script using popen.  I also noticed that each stage in a pipepline created with sh seemed to run fully before passing it's results along (basically didn't handle SIGPIPE).
# Here are some examples:
### Using python sh

```
# sh1.py
from sh import seq 
from sh import gsort
gsort(seq('100000'), '-R')
```
#### time

```
> time ./sh1.py > /dev/null                                                                                                                        
> ./sh1.py > /dev/null  8.17s user 0.75s system 111% cpu 8.033 total
```
### Using good ol' bash

```
# shell1.sh
seq 100000 | gsort -R > /dev/null
```
#### time

```
> time ./shell1.sh > /dev/null                                                                                                                   
> ./shell1.sh > /dev/null  3.62s user 0.07s system 97% cpu 3.788 total
```
### Using python popen

```
# popen1.py
from subprocess import Popen
from subprocess import PIPE
seq = Popen(['seq','100000'], stdout=PIPE)
sort = Popen(['gsort','-R'], stdin=seq.stdout)
sort.wait()
```
#### time

```
> time ./popen1.py > /dev/null                                                                                                               
> ./popen1.py > /dev/null  3.65s user 0.09s system 97% cpu 3.837 total
```
## Testing seq into awk

Things got pretty bad here...
### Using python sh

```
# sh2.py
from sh import seq 
from sh import awk
print awk(seq('100000'), '{print $0 + 1;}')
```
#### time

```
> time ./sh2.py > /dev/null                                                                                                                    
> ./sh2.py > /dev/null 9.33s user 4.51s system 118% cpu 11.648 total
```
### Using good ol' bash

```
# shell2.sh
seq 100000 | awk '{print $0 + 1}'
```
#### time

```
> time ./shell2.sh                                                                                                                        
> ./shell2.sh > /dev/null  0.22s user 0.01s system 130% cpu 0.180 total
```
### Using python popen

```
# popen2.py
from subprocess import Popen
from subprocess import PIPE
seq = Popen(['seq','100000'], stdout=PIPE)
add = Popen(['awk','{print $0 + 1}'], stdin=seq.stdout)
add.wait()
```
#### time

```
> time ./popen2.py > /dev/null                                                                                                                     
> ./popen2.py > /dev/null  0.26s user 0.03s system 122% cpu 0.240 total
```
## Finally - Streaming Behavior

Running any of these commands into head, shows whether or not the process is streaming data or if it's computing everything in each step before passing it along
### Using python sh

Still takes the same amount of time

```
time ./sh2.py| head > /dev/null  0.00s user 0.00s system 0% cpu 10.662 total
```
### Using the other 2 returns instantly

```
time ./popen2.py| head > /dev/null  0.00s user 0.00s system 5% cpu 0.062 total
time ./shell2.sh | head > /dev/null  0.00s user 0.00s system 14% cpu 0.025 total
```
# Summary

As you can see, there's a lot of overhead happening in sh.py compared to bash or using popen.  I'm constantly working with > 1GB files, so that makes sh intractable for me right now.  Are these issues on the roadmap, or is this not really in the scope of sh's concerns?  

PS - Have you considered using dot syntax to build up the pipeline in addition to function composition?  Where something like this would work:

```
cat('hello.txt').sort('-t,', '-k2,2').head()
```

Just think that'd be cool, and more familiar to bashers...  :)

**Comment by amoffat** [comment]
In this example:

```python
from sh import seq 
from sh import gsort
gsort(seq('100000'), '-R')
```

What's happening under the hood is as follows:

1. seq executes, all of its output is aggregated together
2. gsort is called with the aggregated contents of seq

This is not super ideal for two reasons:

1. The first command must finish before the second can run
2. Data from the first command is bloating memory

This topic is covered in [advanced piping](http://amoffat.github.io/sh/index.html#advanced-piping)

So the advised solution is...

```python
from sh import seq 
from sh import gsort
gsort(seq('100000', _piped=True), '-R')
```

This tells `seq` "hey, we're going to use your output to stream into another command *as it comes in*, so let's set up some machinery to do this."  This keyword argument is necessary because otherwise `seq` has no idea how it will be used.

However, this still has overhead problems.  For people like you, working with large amounts of data, this data does not need to flow into python at all.  The processes should be able to be connected together at a more fundamental level.  Fortunately there is a lesser known option: `_piped="direct"`.  I say "lesser known" because I am a bad documentation keeper, and the docs are a little behind.  But this option is what you want:

```python
from sh import seq 
from sh import gsort
gsort(seq('100000', _piped="direct"), '-R')
```

This will set `gsort`'s stdin fd to be `seq`'s stdout fd.  So the overhead should mostly vanish.  Tell me if that works out for you and we can go from there

**Comment by fivetentaylor** [comment]
Thank you for the quick response!

Yes, that made for a huge improvement, but still ~5x slower on the second command:

    > time ./sh2.py > /dev/null                                                                                                  
    ./sh2.py > /dev/null  1.13s user 0.17s system 129% cpu 0.998 total

    > time ./popen2.py > /dev/null                                                                                               
    ./popen2.py > /dev/null  0.20s user 0.03s system 119% cpu 0.190 total

Also, piping a command into head on the shell ignores SIGPIPE, still taking the same amount of time to run:

    > time ./sh2.py| head > /dev/null
    ./sh2.py| head > /dev/null  0.00s user 0.00s system 0% cpu 0.989 total

Thanks for the help,
Taylor

**Comment by amoffat** [comment]
Hmm..I will need to dig in a bit to find the performance overhead of direct piping.  Until I figure out what's making it run much slower, that will be as good as it gets :\

I'll also need to dig into the SIGPIPE issue.  Thanks for reporting both of these

**Comment by fivetentaylor** [comment]
Thank you!!!

**Comment by mandarvaze** [comment]
I've submitted a pull request #243 to document lesser known option: `_piped="direct"`

**Comment by amoffat** [comment]
I experimented a bit with finding the overhead in direct piping (which is now the default for `_piped=True`, so `_piped="direct"` is no longer needed). The overhead exists is mostly due to sh's flexibility, and spending engineering effort to eek out more performance would complicate the code base substantially.  So wrt more performance, it will stay the way it is and I would recommend using `subprocess.Popen` on code you want to really perform well.

Wrt the piping, I believe that is an issue with Python itself. It is the Python process itself that should handle SIGPIPE, and although child processes (ones launched with `sh`) would inherit the signal handlers, I can't make `sh` correctly assume how to handle SIGPIPE globally, as it will effect code that import `sh`.
